Fully qualify types in REPL generated code

dwijnand opened the pull request:
This is so the Scala REPL can work with `-Yno-imports` and `-Yno-predef`.

The partest run case is a bit of a shotgun approach to seeing if I can get it to fall over in any other way, stealing code from existing repl test cases.  If there's a better approach, please let me know.

I'm happy to iterate.

Review by @retronym.

dwijnand committed:
Fully qualify types in REPL generated code

dwijnand commented:
Also please let me know if there are any more places that need changing for `-Yno-imports` and `-Yno-predef`.

Btw, it's very handy that @scala-jenkins automatically adds pull-requests to the right milestone!

retronym commented:
Did you try `:power` mode?

retronym commented:
```
def preambleHeader = "class %s extends Serializable { "
```

(under `-Yrepl-classbased`)

retronym commented:
```
eval("new javax.script.CompiledScript { def eval(context: javax.script.ScriptContext): Object = { " + script + " }.asInstanceOf[Object]; def getEngine: javax.script.ScriptEngine = engine }")
```

retronym commented:
```
"scala.runtime.ScalaRunTime.replStringOf(%s, %s)".format(x, maxlen)
```

som-snytt commented on test/files/run/repl-no-imports-no-predef.scala:
```diff
@@ -0,0 +1,112 @@
+  override def eval() = {
+    val lines = super.eval
+    val r = """\$\$Lambda.*""".r
+    lines.map(l => r.replaceAllIn(l, "<function0>"))
```
You can override `normalize` https://github.com/scala/scala/blob/2.12.x/src/partest-extras/scala/tools/partest/ReplTest.scala#L24

som-snytt commented:
That's great. I noticed that REPL was dissed on gitter for not supporting `-Yno-nothing`, so I was going to do it, too. I just verified that fixing `_root_.java.io.Serializable` makes `-Yrepl-class-based` work.

I think retronym's comment on CompiledScript is now at

https://github.com/scala/scala/blob/2.12.x/src/repl/scala/tools/nsc/interpreter/Scripted.scala#L45

som-snytt commented:
Boy, that was kind of annoying, but:

https://github.com/som-snytt/scala/commit/faf6059845712bc6f7d2381c8fc67b2fd31cdafd

dwijnand commented:
Cherry-picked @som-snytt's work in `Scripted`, updated to test for `:power`, `-Yrepl-classbased` and switched from `ReplTest#normalise`.

@retronym 
`"scala.runtime.ScalaRunTime.replStringOf(%s, %s)".format(x, maxlen)` I think is fine, the string part is fully qualified, and the `.format` part isn't running under `-Yno-predef` so the string gets augmented.

I can't figure out where

```scala
eval("new javax.script.CompiledScript { def eval(context: javax.script.ScriptContext): Object = { " + script + " }.asInstanceOf[Object]; def getEngine: javax.script.ScriptEngine = engine }")
```

comes from, how to trigger it, or even better how to test for it.

Thank you for the review.

retronym commented:
>  I think is fine, the string part is fully qualified

I wan't able to trigger a failure with it, but what I had in mind was `"_root_.scala.runtime...."`.

dwijnand commented:
Done.

retronym commented:
The change is looking comprehensive now.

I used IntelliJ to search in literal strings for suspicious patterns:

![image](https://cloud.githubusercontent.com/assets/65551/15574724/6ac15fea-2392-11e6-9555-daee0ca0f57c.png)

retronym commented:
`_root_` is an extra safeguard if the generated code is nested within a context that has unknown enclosing imports or defintions, that might shadow `java` or `scala`. It would be good to use this consistently IMO.

dwijnand commented:
Updated.

som-snytt commented:
Yes, I had `_root_`ed but when the unit test still failed, I backed it out, thinking I must have mistyped something; it turned out to be a missing `scala.Any` in the init code for Scripted, and the failure was ignored, because how could it not compile? Too late at night to re-`_root_`, sorry. Feel free to squash my commit.

dwijnand commented:
Thanks @som-snytt, squashed.

adriaanm commented:
🌲 root stuff! 🌲

dwijnand commented:
oh, it's green!

som-snytt commented:
I was rooting for you.

retronym commented:
LGTM. Thanks and welcome, @dwijnand!

som-snytt commented on test/files/run/repl-no-imports-no-predef-classbased.scala:
```diff
@@ -0,0 +1,19 @@
+import c.k
+scala.Predef.implicitly[K]
+val k = 42
+k                 // was K(OK?)
```
I don't know what this is trying to test. There's the bug with `-Yrepl-class-based` that imports of implicits aren't handled correctly, but that has nothing to do with `-Yno-imports`. https://issues.scala-lang.org/browse/SI-9799

dwijnand replied:
As I tried to convey in the pull request description I took a bit of a shotgun approach to trying to find edge cases to fix, stealing code from existing repl test cases.

som-snytt replied:
OK, thanks. The test fails a couple of ways on #5220 which I'm trying to put to bed before a flight later...
